apps/law_memory/pipeline.py

```
@@ -0,0 +1,98 @@
+from .visual.diagnostics import plot_entropy_loop, plot_law_attractor
+from .chaos import compute_torsion_spectrum
+from .metrics import inject_into_metrics
+from .observables import density_to_gamma
+
+
+def _load_density(path: Path) -> np.ndarray:
+    data = np.load(path)
+    if data.ndim == 1:
+        return data[np.newaxis, :]
+    if data.ndim != 2:
+        raise ValueError("density input must be 1D or 2D")
```

Ruff (0.13.3)

Avoid specifying long messages outside the exception class

(TRY003)